pkg/await/apps_deployment.go

```diff
@@ -0,0 +1,385 @@
+			period <- struct{}{}
+		}
+	}()
+
+	return dia.await(deploymentWatcher, podWatcher, time.Tick(5*time.Minute), period)
+}
+
+// await is a helper companion to `Await` designed to make it easy to test this module.
+func (dia *deploymentInitAwaiter) await(
+	deploymentWatcher, podWatcher watch.Interface, timeout <-chan time.Time, period chan struct{},
+) error {
+	inputPodName := dia.config.currentInputs.GetName()
+	for {
+		// Check whether we've succeeded.
+		if dia.rolloutComplete {
+			return nil
+		}
+
+		// Else, wait for updates.
+		select {
+		case <-dia.config.ctx.Done():
+			// On cancel, check one last time if the service is ready.
+			if dia.rolloutComplete {
+				return nil
+			}
+			return &cancellationError{
+				objectName: inputPodName,
+				subErrors:  dia.errorMessages(),
+			}
+		case <-timeout:
+			// On timeout, check one last time if the service is ready.
+			if dia.rolloutComplete {
+				return nil
+			}
+			return &timeoutError{
+				objectName: inputPodName,
+				subErrors:  dia.errorMessages(),
+			}
+		case <-period:
+			scheduleErrors, containerErrors := dia.aggregatePodErrors()
+			for _, message := range scheduleErrors {
+				dia.warn(message)
+			}
+
+			for _, message := range containerErrors {
+				dia.warn(message)
+			}
+		case event := <-deploymentWatcher.ResultChan():
+			dia.processDeploymentEvent(event)
+		case event := <-podWatcher.ResultChan():
+			dia.processPodEvent(event)
+		}
+	}
+}
+
+func (dia *deploymentInitAwaiter) processDeploymentEvent(event watch.Event) {
+	// Start with a blank slate.
+	dia.rolloutComplete = false
+	dia.deploymentErrors = map[string]string{}
+
+	inputDeploymentName := dia.config.currentInputs.GetName()
+
+	deployment, isUnstructured := event.Object.(*unstructured.Unstructured)
+	if !isUnstructured {
+		glog.V(3).Infof("Deployment watch received unknown object type '%s'",
+			reflect.TypeOf(deployment))
+		return
+	}
+
+	// Do nothing if this is not the Deployment we're waiting for.
+	if deployment.GetName() != inputDeploymentName {
+		return
+	}
+
+	// Mark the rollout as incomplete if it's deleted.
+	if event.Type == watch.Deleted {
+		return
+	}
+
+	// Get current generation of the Deployment.
+	dia.currentGeneration = deployment.GetGeneration()
+	if dia.currentGeneration == 0 {
+		// No current generation, Deployment controller has not yet created a ReplicaSet. Do
+		// nothing.
+		return
+	}
+
+	// Get observed generation of the Deployment.
+	rawGeneration, hasGeneration := openapi.Pluck(deployment.Object, "status", "observedGeneration")
+	var isNumber bool
+	dia.observedGeneration, isNumber = rawGeneration.(int64)
+	if !hasGeneration || !isNumber {
+		// No observed generation, Deployment controller has not yet created a ReplicaSet. Do
+		// nothing.
+		dia.observedGeneration = 0
+		return
+	}
+
+	// Return if Deployment controller has not created a new ReplicaSet for `currentGeneration`.
+	if dia.observedGeneration != dia.currentGeneration {
+		return
```

Comment:
If we don't get to the `currentGeneration`, we should fail. A place this could happen is if they run `kubectl rollback`, in which case `.metadata.generation` would be set to the previous generation, and `observedGeneration` would need to be _decremented_.